bitcoin/src/blockdata/transaction.rs

```diff
@@ -1147,6 +1161,107 @@ impl From<&Transaction> for Wtxid {
     }
 }
 
+/// Predicts the weight of to-be-constructed transaction.
+///
+/// This function computes the weight of a transaction which is not fully known. All that is needed
+/// is the lengths of scripts and witness elements.
+///
+/// # Arguments
+///
+/// * `inputs` - iterator which returns `InputWeightPrediction` for each input of to-be-constructed
+///   transaction.
+/// * `output_script_lens` - iterator which returns the length of `script_pubkey` of each input of
+///   to-be-constructed transaction.
+///
+/// # Usage
+///
+/// When signing a transaction one doesn't know the signature before knowing transaction fee and
+/// transaction fee is not known before knowing transaction size which is no known before knowing
+/// signature. This apparent dependency cycle can be broken by knowing the length of the signature
+/// without knowing the contents of the signature. E.g. we know all Schnorr signatures are 64 bytes
+/// long.
+///
+/// Additionally, some protocols may require calculating the amounts before knowing various parts
+/// of the transaction (assuming their length is known).
+///
+/// # Notes on integer overflow
+///
+/// Overflows are intentionally not checked because either of these is true:
+///
+/// * The transaction is valid (obey the block size limit) and the code feeds correct values to
+///   this function - no overflow can happen.
+/// * The transaction will be so large it doesn't fit in the memory - the overflow will happen but
+///   then the transaction will fail to construct and even if one serialized it on disk directly
+///   it'd be invalid anyway so overflow doesn't matter.
+/// * The values fed into this function are inconsistent with the actual lengths the transaction
+///   will have - the code is already broken and checking overflows doesn't help. Unfortunately
+///   this probably cannot be avoided.
```

Comment:
I read this closely to understand everything, here are a few edits. This in no way implies a claim to perfect grammar, I just did my best in an effort to produce the best docs possible for our project.
```
/// Predicts the weight of a to-be-constructed transaction.
///
/// This function computes the weight of a transaction which is not fully known. All that is needed
/// is the lengths of scripts and witness elements.
///
/// # Arguments
///
/// * `inputs` - an iterator which returns `InputWeightPrediction` for each input of the
///   to-be-constructed transaction.
/// * `output_script_lens` - an iterator which returns the length of `script_pubkey` of each input
///   of the to-be-constructed transaction.
///
/// # Usage
///
/// When signing a transaction one doesn't know the signature before knowing the transaction fee and
/// the transaction fee is not known before knowing the transaction size which is not known before
/// knowing the signature. This apparent dependency cycle can be broken by knowing the length of the
/// signature without knowing the contents of the signature e.g., we know all Schnorr signatures
/// are 64 bytes long.
///
/// Additionally, some protocols may require calculating the amounts before knowing various parts
/// of the transaction (assuming their length is known).
///
/// # Notes on integer overflow
///
/// Overflows are intentionally not checked because one of the following holds:
///
/// * The transaction is valid (obeys the block size limit) and the code feeds correct values to
///   this function - no overflow can happen.
/// * The transaction will be so large it doesn't fit in the memory - overflow will happen but
///   then the transaction will fail to construct and even if one serialized it on disk directly
///   it'd be invalid anyway so overflow doesn't matter.
/// * The values fed into this function are inconsistent with the actual lengths the transaction
///   will have - the code is already broken and checking overflows doesn't help. Unfortunately
///   this probably cannot be avoided.
```